src/handlers/merge_conflict.rs

```diff
@@ -0,0 +1,330 @@
+    let (conflicting, unknowns): (Vec<_>, Vec<_>) = prs
+        .into_iter()
+        .filter(|pr| pr.mergeable != MergeableState::Mergeable)
+        // Assume that pushes to other branches won't affect this PR (maybe
+        // not the greatest assumption, but might help with some noise). In
+        // practice, this shouldn't matter much since simultaneous pushes to
+        // multiple branches is rare.
+        .filter(|pr| pr.base_ref_name == branch_name)
+        .partition(|pr| pr.mergeable == MergeableState::Conflicting);
+
+    for conflict in conflicting {
+        let issue = repo.get_issue(gh, conflict.number).await?;
+        maybe_add_comment(gh, &mut db, config, &issue, possibly.as_deref()).await?;
+    }
+    if !unknowns.is_empty() {
+        let config = config.clone();
+        tokio::task::spawn(async move {
+            // See module note about locking.
+            tokio::time::sleep(UNKNOWN_RESCAN_DELAY).await;
+            let gh = GithubClient::new_from_env();
+            // NOTE: The `possibly` here is even less likely to be correct due
+            // to the risk that another push happened while we were waiting.
+            // May want to consider changing it to `None` if it regularly
+            // points to the wrong thing.
+            if let Err(e) = scan_unknowns(&gh, db, &config, &repo, &unknowns, possibly).await {
+                log::error!("failed to scan unknown PRs for merge conflicts: {e:?}");
+            }
+        });
+    }
+
+    Ok(())
+}
+
+/// Scans open PRs with an unknown mergeable status to see if the mergeability
+/// has been updated.
+async fn scan_unknowns(
+    gh: &GithubClient,
+    mut db: PooledClient,
+    config: &MergeConflictConfig,
+    repo: &Repository,
+    unknowns: &[MergeConflictInfo],
+    possibly: Option<String>,
+) -> anyhow::Result<()> {
+    log::debug!(
+        "re-scanning {} unknown PRs for merge conflicts for {}",
+        unknowns.len(),
+        repo.full_name
+    );
+    for unknown in unknowns {
+        let issue = repo.get_issue(&gh, unknown.number).await?;
+        // Ignore None, we don't want to repeatedly hammer GitHub asking for the answer.
+        if issue.mergeable == Some(false) {
+            maybe_add_comment(gh, &mut db, config, &issue, possibly.as_deref()).await?;
+        } else if issue.mergeable == None {
+            log::info!("unable to determine mergeable after delay for {unknown:?}");
+        }
+    }
+    Ok(())
+}
+
+async fn maybe_add_comment(
+    gh: &GithubClient,
+    db: &mut DbClient,
+    config: &MergeConflictConfig,
+    issue: &Issue,
+    possibly: Option<&str>,
+) -> anyhow::Result<()> {
+    // Note: This database access is racey. Preferably there would be a
+    // row-lock held between the time it is loaded and the save call below.
+    // The `post_comment` call should normally be pretty fast, so races should
+    // be rare. If they happen too often, consider adding some locking
+    // mechanism here.
```

Comment:
I think there *is* a lock held, actually on the full table? `load(...)` starts with `LOCK TABLE issue_data` and presumably that's kept until the transaction ends (which happens in save).

https://github.com/rust-lang/triagebot/blob/742b66bf475296212ebd913cfa909b1dbc747973/src/db/issue_data.rs#L41

Do we think we need more?

Reply:
I don't remember what I was thinking at the time. I saw the LOCK TABLE call, but for some reason was thinking there was something else. I can't think of what it is now, and the lock is clearly held in the `IssueData`, so I removed the comment. Thanks for pointing that out!